api_test.go

```diff
@@ -220,61 +221,100 @@ func TestAPIReportEvent(t *testing.T) {
-func createTempPasswordFile(filename, password string) error {
-	f, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE, 0775)
-	f.Truncate(0)
-	f.Seek(0, 0)
-
-	defer f.Close()
-	if err != nil {
-		return err
-	}
-	_, err = f.WriteString(password)
-	return err
-}
-
-func readPassword(filename string) (string, error) {
-	content, err := ioutil.ReadFile(filename)
-	return string(content), err
-}
-
-func getLock(filename string) (*flock.Flock, bool, error) {
-	fileLock := flock.New(tempPasswordFile)
-	lockCtx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
-	defer cancel()
-	locked, err := fileLock.TryLockContext(lockCtx, 678*time.Millisecond)
-	return fileLock, locked, err
-}
-
 func TestPasswordLock(t *testing.T) {
 	tempPassword := randString(20)
-	err := createTempPasswordFile(tempPasswordFile, tempPassword)
-	require.Equal(t, err, nil, "must be able to create/open "+tempPasswordFile)
+	confPassword := NewConfigPassword(tempPasswordFile)
+	anotherConfPassword := NewConfigPassword(tempPasswordFile)
+
+	err := confPassword.WritePassword(tempPassword)
+	assert.NotEqual(t, nil, err)
 
-	fileLock, locked, err := getLock(tempPasswordFile)
+	locked, err := confPassword.GetExLock()
+	defer confPassword.Unlock()
 	require.True(t, locked, "File lock must succeed")
-	require.Equal(t, err, nil, "File lock must succeed")
+	require.Equal(t, nil, err, "must be able to get lock "+tempPasswordFile)
+
+	err = confPassword.WritePassword(tempPassword)
+	require.Equal(t, nil, err, "must be able to write to"+tempPasswordFile)
 
-	err = WritePassword(tempPasswordFile, randString(20))
+	locked, err = anotherConfPassword.GetExLock()
 	assert.NotEqual(t, nil, err)
-	fileLock.Unlock()
+	assert.False(t, locked)
 
-	currentPassword, err := readPassword(tempPasswordFile)
-	assert.Equal(t, err, nil)
+	err = anotherConfPassword.WritePassword(randString(20))
+	assert.NotEqual(t, nil, err)
+	confPassword.Unlock()
+
+	currentPassword, err := confPassword.ReadPassword()
+	assert.Equal(t, nil, err)
 	assert.Equal(t, tempPassword, currentPassword)
 
 	tempPassword = randString(20)
-	err = WritePassword(tempPasswordFile, tempPassword)
-	assert.Equal(t, err, nil)
+	locked, err = anotherConfPassword.GetExLock()
+	defer anotherConfPassword.Unlock()
+	assert.Equal(t, nil, err)
+	assert.True(t, locked)
 
-	currentPassword, err = readPassword(tempPasswordFile)
-	assert.Equal(t, err, nil)
-	assert.NotEqual(t, currentPassword, tempPassword)
+	err = anotherConfPassword.WritePassword(tempPassword)
+	assert.Equal(t, nil, err)
+
+	currentPassword, err = anotherConfPassword.ReadPassword()
+	assert.Equal(t, nil, err)
+	assert.Equal(t, tempPassword, currentPassword)
 
 	err = os.Remove(tempPasswordFile)
 }
 
+func createTestConfig() error {
+	conf := &Config{
+		ServerURL:  apiURL,
+		Group:      defaultGroup,
+		DeviceName: randString(10),
+	}
+	d, err := yaml.Marshal(conf)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(testConfig, d, 0600)
+	return err
+}
```

Comment:
A useful pattern for test helpers like this is for them to take the `*testing.T` so that the can helper can use `require.NoError` etc to deal with anything that goes wrong. This means the caller doesn't have to worry about error handling, making the tests for concise and readable

Reply:
Another useful pattern here would be to return a function which will clean up the config file that was created. This can be used in a defer by the calling test. `removeTestConfig` wouldn't be needed then.